File: java/MiniDuckSimulator-StrategyPattern/src/dev/dmohindru/Duck.java
package dev.dmohindru;

public abstract class Duck {
    protected FlyBehavior flyBehavior;
    protected QuackBehavior quackBehavior;
    public Duck() {}

    public abstract void display();

    public void performFly() {
        flyBehavior.fly();
    }

    public void performQuack() {
        quackBehavior.quack();
    }

    public void setFlyBehavior(FlyBehavior fb) {
        this.flyBehavior = fb;
    }

    public void setQuackBehavior(QuackBehavior qb) {
        this.quackBehavior = qb;
    }

    public void swim() {
        System.out.println("All ducks float, even decoys!");
    }
}

File: nodejs/MiniDuckSimulator/src/duck.ts
import {FlyBehavior} from "./flybehavior";
import {QuackBehavior} from "./quackbehavior";

export abstract class Duck {
    constructor(protected _flyBehavior: FlyBehavior, protected _quackBehavior: QuackBehavior) {
    }

    abstract display(): void;

    performFly(): void {
        this._flyBehavior.fly();
    }

    performQuack(): void {
        this._quackBehavior.quack();
    }

    set flyBehavior(fb: FlyBehavior) {
        this._flyBehavior = fb;
    }

    set quackBehavior(qb: QuackBehavior) {
        this._quackBehavior = qb;
    }

    swim(): void {
        console.log("All ducks float, even decoys!");
    }
}